Relax requirement for `glyphs` when using `text-field`

1ec5 opened the pull request:
Adjusted the `glyphs` documentation to relax the informal requirement for this property when `text-field` occurs in a symbol layer. The documentation now matches the machine-readable specification in classifying `glyphs` as optional. Also clarified the interpretation of `text-font` when `glyphs` is set versus unset. When it’s unset, the array is interpreted as a cascading fallback list. This is how it already works on each platform when setting the ideographic font, but now it‘s explicit and coupled to another part of the style specification.

Fixes #1045.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [ ] ~~Include before/after visuals or gifs if this PR includes visual changes.~~
 - [ ] ~~Write tests for all new functionality.~~
 - [x] Document any changes to public APIs.
 - [ ] ~~Post benchmark scores.~~
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented:
LGTM. Waiting for @louwers review as well.

louwers commented:
MapLibre Native seems to break down pretty bad when `glyphs` is missing.

<img width="1024" alt="image" src="https://github.com/user-attachments/assets/a015249a-2513-4025-8318-338ef04db5b3" />

No graceful degradation like MapLibre GL JS. 🫤

<img width="1544" alt="image" src="https://github.com/user-attachments/assets/fdf58b0b-8cf5-47ab-8045-28099b5a222f" />

louwers reviewed (changes requested):
I am not in favor of this change, I think it is better to always require people to set `glyphs` (even if it is just a fallback font) and add APIs to override the glyphs property with custom font loading. This offers much better backward compatibility for future styles.

That said, I don't want to be the sole blocker if other people don't have an issue with this. That said there is something that I think we should definitely change before merging this:

We document what versions of MapLibre GL JS and MapLibre Native support which features, and which behavior. We should add a support matrix to this entry with 'optional when using fonts'. There should be a ❌ for Native along with a tracking issue, to make people aware that when they author styles without a `glyphs` property, all bets are off if that style is used with MapLibre Native (see comment above). Of course, users of older MapLibre GL JS versions need to understand too what version of MapLibre GL JS they need to use to take advantage of 'font loading from the environment'.

---

Again, I believe that font loading should be a responsibility of the style author. A style should incorporate everything that is needed to show the map. If the PBF format is too limited, we should devise and agree upon another cross-platform font format. Hand waving and just telling people to make sure fonts are available in 'the environment' is inconvenient for users and may split the ecosystem.

1ec5 commented:
> MapLibre Native seems to break down pretty bad when glyphs is missing.

Would you accept a PR to MapLibre Native that implements maplibre/maplibre-gl-js#4564?

> If the PBF format is too limited, we should devise and agree upon another cross-platform font format. Hand waving and just telling people to make sure fonts are available in 'the environment' is inconvenient for users and may split the ecosystem.

If locally rendered fonts are problematic, then they’re already problematic with respect to CJK, because the developer only needs to name a font “in the environment”, if anything. The named font doesn’t have to be bundled and doesn’t even have to exist; the implementation falls back to a system font on both [iOS](https://github.com/maplibre/maplibre-native/blob/3fc93a0b024e34514dafcbb424db93593ff540be/platform/darwin/core/local_glyph_rasterizer.mm#L111-L117) and [Android](https://github.com/maplibre/maplibre-native/blob/3fc93a0b024e34514dafcbb424db93593ff540be/platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/text/LocalGlyphRasterizer.java#L54).[^android] Locally rendered fonts have been enabled by default for CJK since GL JS v0.0.16, Android v5.3.0, and iOS v3.7.3. Any ecosystem split should’ve been apparent among CJK-serving developers by now.

> We document what versions of MapLibre GL JS and MapLibre Native support which features, and which behavior. We should add a support matrix to this entry with 'optional when using fonts'. There should be a ❌ for Native along with a tracking issue, to make people aware that when they author styles without a glyphs property, all bets are off if that style is used with MapLibre Native (see comment above). Of course, users of older MapLibre GL JS versions need to understand too what version of MapLibre GL JS they need to use to take advantage of 'font loading from the environment'.

Sure, I’ve added a compatibility table for `glyphs` and added to the existing compatibility table for `text-font`, linking to maplibre/maplibre-gl-js#3302 and maplibre/maplibre-native#165.

[^android]: Based on the behavior of [`Typeface.create()`](https://developer.android.com/reference/android/graphics/Typeface#create(android.graphics.Typeface,%20int)).

louwers commented:
Thanks! Looks like we still need to add some code to make sure the table renders (or add a table to the Markdown string):

<img width="1541" alt="image" src="https://github.com/user-attachments/assets/69665a06-e476-428f-b94c-3cacbcf4dff2" />

> Locally rendered fonts have been enabled by default for CJK 

Sure, but it's a bit of a hack the way it's implemented, I think you agree. That's my whole point, font faces are an important part of the look and feel of maps. If we're going to rely on the environment for Western fonts too, we will have even more cases where the best we will be able to do is use a fallback font instead of the font intended by the style author.

> Would you accept a PR to MapLibre Native that implements https://github.com/maplibre/maplibre-gl-js/pull/4564?

Yes, but what font would you use?

What I would really like is if we could use web fonts with MapLibre Native too (and embed web fonts directly in a style). Something like [this](https://developer.chrome.com/blog/memory-safety-fonts) looks promising (we can compile [Rust code](https://maplibre.org/maplibre-native/docs/book/rust.html) into MapLibre Native now).

1ec5 commented:
> Looks like we still need to add some code to make sure the table renders

I ran `WATCH=1 npm run generate-docs` per the readme and I’m seeing both compatibility tables. It should work just like for the `centerAltitude` root property.

![SDK Support table under the glyphs heading](https://github.com/user-attachments/assets/2c8c2572-118a-4828-8921-53851965f54a)

![SDK Support table under the text-font heading, with an extra row for local fonts](https://github.com/user-attachments/assets/fa669dd5-fee2-4c86-b77e-1aa08cd69ad1)

The command didn’t generate anything else for me to check in. Can you try again locally?

>> Locally rendered fonts have been enabled by default for CJK
> 
> Sure, but it's a bit of a hack the way it's implemented, I think you agree. That's my whole point, font faces are an important part of the look and feel of maps. If we're going to rely on the environment for Western fonts too, we will have even more cases where the best we will be able to do is use a fallback font instead of the font intended by the style author.

It’s a matter of degree. Just yesterday, I was explaining the current text rendering system to some of the folks who maintain the AARoads Wiki, which proudly uses MapLibre ([example](https://wiki.aaroads.com/wiki/A1_(Uganda))). They were in disbelief that it relies on spritesheets and glyph atlases, which they regard as a video game hack from the ’80s. Rest assured, I’m not seeking to revamp WebGL’s native text drawing functionality. 😅

I agree that working directly with actual font data and a fit-for-purpose HarfBuzz would be superior to, essentially, trafficking in screenshots of text. I will make sure anything I do in this area facilitates rather than impeding work towards that goal. But in my view, the change I’m proposing is essential for both local font and complex text support in MapLibre. Complex text support requires us to change how we segment the text, so that each glyph in the atlas represents a string of arbitrary length rather than a single character. This is fundamentally incompatible with the existing glyph PBF format.

Going forward, we can let the style author choose between traditional glyph PBFs and “handwavy” fonts from the local environment, based on their priorities. But if we insist on relegating local fonts to a runtime override, style authors will have to juggle two different interpretations of `text-font` simultaneously. Practically speaking, the local option means they’ll include a generous selection of font names, to catch as many available fonts as possible on each platform, which will be wildly incompatible with the traditional comma-separated fontstack interpretation. We shouldn’t force the style author to upload a hideous set of Code2000 glyphs with a fontstack name of `Euphemia%20UCAS,Euphemia,Aboriginal%20Sans,OskiBlackfoot,Code2000,Everson%20Mono%20Unicode`,[^wiktionary] just to align it to a reasonably comprehensive local fallback list for indigenous Canadian place names.

If I may, MapLibre Native has been on a journey to support complex writing systems the right way for over two years, with little to show to end users so far. Meanwhile, I’ve demonstrated that extended SDF can address that need at least functionally. I support the HarfBuzz effort and want it to succeed, but I also think developers and users should benefit from something tangible in the meantime. In seeking visual perfection, we should not lose sight of the fact that a beautiful, consistent font is useless if the text is missing and potentially offensive if the text is left garbled.

>> Would you accept a PR to MapLibre Native that implements https://github.com/maplibre/maplibre-gl-js/pull/4564?
> 
> Yes, but what font would you use?

I’m offering to implement the full feature in both repositories, without a hard-coded fallback font. This means the style author would choose the fonts based on what they provide or what’s likely to be available on the platform, as they’re already used to doing. I had previously floated the idea of a hard-coded fallback to buy myself some time before committing to the native implementation, but at this point there’s little point in holding off on a full implementation. As a practical matter, I probably would still stage complex text segmentation in GL JS first, but that shouldn’t be an interoperability concern.

> What I would really like is if we could use web fonts with MapLibre Native too (optionally also embed web fonts directly in a style). Something like [this](https://developer.chrome.com/blog/memory-safety-fonts) looks promising (we can compile [Rust code](https://maplibre.org/maplibre-native/docs/book/rust.html) into MapLibre Native now). Because parity between MapLibre GL JS and MapLibre Native in terms of how a particular style looks is one of our aims.

A typical MapLibre-powered application consists of more than a map. Very often, a developer needs to apply a corporate font across all three platforms. The Web almost always uses a different font asset than the mobile platforms, distributed through different means, even for a visually identical typeface. There are both technical and legal reasons for this, as I pointed out in https://github.com/maplibre/maplibre-style-spec/issues/868#issuecomment-2721877794. This isn’t a new problem, so let’s take advantage of the solutions that developers are already using for their applications.

We can explore Web fonts for the native platforms separately. If extended SDF ships on Android or iOS without Web font support, it just means developers who are already adequately served by glyph PBFs will stick to glyph PBFs, while those who need complex text will settle for system fonts as a last resort.

[^wiktionary]: [More where that came from](https://en.wiktionary.org/wiki/MediaWiki:Gadget-LanguagesAndScripts.css).

louwers commented:
Thanks for taking the time to so argue this change so extensively.

> Can you try again locally?

Ah right, it shows up in the 'root properties' page, but I was looking at http://localhost:8000/maplibre-style-spec/glyphs/

I will have a look how we can add it there too.

> They were in disbelief that it relies on spritesheets and glyph atlases, which they regard as a video game hack from the ’80s. 

Lol. To be fair, they probably weren't using SDFs for glyphs in the 80's?

> I’m offering to implement the full feature in both repositories, without a hard-coded fallback font.

That would be great. I still believe that font distribution should be part of styles to make sure works reliably everywhere. There is quite a lot of variability in terms of what fonts are available in the Native world. But that can still be addressed with other style spec proposals.

1ec5 commented:
Thanks for fixing the other page; didn’t realize it was there too.

#868 does make sense from a certain perspective. A webpage is to a CSS stylesheet defining `@font-face` as a map in the webpage is to a MapLibre stylesheet defining font locations. In the native Web stack, font substitution makes this tenable. Hopefully we can do both, in which case the self-contained fonts will take precedence.
